analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -29,5 +29,81 @@ public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
+            && IsPropertyAccessorAccessibleFromOtherAssemblies(accessor.Modifiers);
+
+        static bool IsAccessibleFromOtherAssemblies(SyntaxNode node) =>
+            node.AncestorsAndSelf().OfType<MemberDeclarationSyntax>().FirstOrDefault() is { } containingMember
+            && node.Ancestors().OfType<BaseTypeDeclarationSyntax>().FirstOrDefault() is { } containingType
+            && IsMemberAccessibleFromOtherAssemblies(containingMember.Modifiers(), containingType)
+            && IsTypeAccessibleFromOtherAssemblies(containingType.Modifiers);
+
+        static bool IsMemberAccessibleFromOtherAssemblies(SyntaxTokenList modifiers, BaseTypeDeclarationSyntax containingType) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || (containingType is InterfaceDeclarationSyntax && HasNoDeclaredAccessabilityModifier(modifiers));
+
+        static bool IsPropertyAccessorAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || HasNoDeclaredAccessabilityModifier(modifiers);
+
+        static bool IsTypeAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers);
+
+        static bool IsPublicOrProtectedOrProtectedInternal(SyntaxTokenList modifiers) =>
+            modifiers.AnyOfKind(SyntaxKind.PublicKeyword)
+            || (modifiers.AnyOfKind(SyntaxKind.ProtectedKeyword) && !modifiers.AnyOfKind(SyntaxKind.PrivateKeyword));
+
+        static bool HasNoDeclaredAccessabilityModifier(SyntaxTokenList modifiers) =>
+            !modifiers.Any(x => x.IsAnyKind(
+                SyntaxKind.PrivateKeyword,
+                SyntaxKind.ProtectedKeyword,
+                SyntaxKind.InternalKeyword,
+                SyntaxKind.PublicKeyword));
+
+        static bool MethodDereferencesArguments(BaseMethodDeclarationSyntax method)
+        {
+            var argumentNames = method.ParameterList.Parameters.Select(x => x.GetName()).ToArray();
+            var walker = new ArgumentDereferenceWalker(argumentNames);
+            walker.SafeVisit(method);
+            return walker.DereferencesMethodArguments;
+        }
+    }
+
+    private sealed class ArgumentDereferenceWalker : SafeCSharpSyntaxWalker
+    {
+        private readonly string[] argumentNames;
+
+        public bool DereferencesMethodArguments { get; private set; }
+
+        public ArgumentDereferenceWalker(string[] argumentNames) =>
+            this.argumentNames = argumentNames;
+
+        public override void Visit(SyntaxNode node)
+        {
+            if (!DereferencesMethodArguments)
+            {
+                base.Visit(node);
+            }
+        }
+
+        public override void VisitIdentifierName(IdentifierNameSyntax node) =>
+            DereferencesMethodArguments |=
+                argumentNames.Contains(node.GetName())
+                && node.Ancestors().Any(x => x.IsAnyKind(
```

Comment:
Shouldn't it be enough to check the parent only? What do you consider a dereference? I would say
* `p.Invocation();` but not `p?.Invocation();`
* `p.PropertyAccess`, but not `p?.PropertyAccess;`
* `p[0];` but not `p?[0];`
* `p->Deref();` // Unsafe code only. Can p null here null? I don't know. I think p must be a struct but I'm not sure.
* `await p;` // FYI: there is an `await? p` language proposal also https://github.com/dotnet/csharplang/issues/35
* `foreach(var i in p)`
* `throw p`
* `p` + `p` No. I think all unary and binary operators like + or ! are okay
* `M(p)` No. Passing around is okay.

Is there something else?
I think in all cases we just need to look at the parent.

Reply:
This is more relevant for VB as there is no casing difference, but we might also add another check, where we also check, that the identifier is indeed referring to a parameter:
In a case like so:
```cs
void M(int Length)
{
    someList.Length.ToString(); // Length can not be a parameter here
}
```

Reply:
I changed the syntax walker to check the parents instead of all ancestors.
I'm not sure about the additional check. Is there an easy way to do it without using the semantic model?

Reply:
Please add a test case for a [PointerMemberAccess](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwUX4DMABEYwMIUDeFjPzTAzgjQBjBGwD245Bn6duvBo1hiAGgG55PRcsYBNDeV59G4gG4BTZMigZzzKsUYAVcQGUhsAOYAKAJSMAXgA+RgASACJvDhUAXzhGDl0Y33CDGM1jfCoANkZ0fggAMzt8ABZGAFloGD9eDK5DI14K7xhUABt23wMjdMaeDMUs3PyikvKW1klpfgAqRgAHX3qMowWAWiCVQMZ6HqaeDaDdHdL9g6yATm8jl3drGB9fbp4AeleTVAQFr5BGb3o8VKy36jBiRgofSAA==)

@pavel-mikula-sonarsource Do we want to support dereferences in unsafe code like in the linked example?

Reply:
> I'm not sure about the additional check. Is there an easy way to do it without using the semantic model?

You just need to make sure, that there is nothing left to the parameter name:
* parameter.Invoke() // <- okay
* somethingLeft.parameter.Invoke() // <- not okay

@Tim-Pohlmann added `GetLeftOfDot` or something like that recently. You can use that to get the left-hand side of what you have found as a candidate.

Reply:
I tried to find `GetLeftOfDot` and similar methods in Tim's PR, but couldn't find them. Any chance you remember where it is?
Does it take more complex expressions into consideration? Like:
`(parameter as object).ToString(); // Noncompliant`

Reply:
It's not implemented yet: https://github.com/SonarSource/sonar-dotnet/issues/6863

Reply:
`GetLeftOfDot` is in Roslyn, but we don't have it in our code base yet. I have used a workaround and created an issue to add it. Maybe this helps you:
https://github.com/SonarSource/sonar-dotnet/pull/6753#discussion_r1126413752

Reply:
As it's not yet available in the project, I'll address this in another PR.